### ISSUE TITLE
Construção do projeto

### ISSUE THREAD

**Issue by LucasBiason** [LucasBiason/ia-ml-knowledge-base#1]
# 🔄 AUTOMATED-ML-PIPELINE

[![Python](https://img.shields.io/badge/Python-3.9+-blue.svg)](https://www.python.org/downloads/)
[![FastAPI](https://img.shields.io/badge/FastAPI-0.104+-green.svg)](https://fastapi.tiangolo.com/)
[![MLflow](https://img.shields.io/badge/MLflow-2.8+-orange.svg)](https://mlflow.org/)
[![DVC](https://img.shields.io/badge/DVC-3.0+-purple.svg)](https://dvc.org/)
[![LangChain](https://img.shields.io/badge/LangChain-0.1+-yellow.svg)](https://langchain.com/)
[![Docker](https://img.shields.io/badge/Docker-20.10+-blue.svg)](https://www.docker.com/)
[![License](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)
[![Status](https://img.shields.io/badge/Status-Planejado-gray.svg)](https://github.com/LucasBiason/ia-ml-knowledge-base)
[![Tests](https://img.shields.io/badge/Tests-pytest-green.svg)](https://docs.pytest.org/)

> **Pipeline Automatizado com MLOps Completo**  
> Sistema profissional para experimentação, versionamento e deploy automatizado de modelos de Machine Learning com integração de LLMs.

---

## 🎯 **OBJETIVO DO PROJETO**

O **Automated-ML-Pipeline** é o terceiro projeto do portfólio IA Knowledge Base, focado em criar um pipeline MLOps completo com:

- 🔬 **Experiment Tracking** - Rastreamento completo de experimentos com MLflow
- 📊 **Data Versioning** - Versionamento de dados com DVC
- 🤖 **Fine-tuning** - Ajuste fino de modelos com LLMs
- 🔍 **RAG Systems** - Retrieval-Augmented Generation
- 🚀 **CI/CD for ML** - Deploy automatizado de modelos
- 📈 **Model Monitoring** - Monitoramento de performance e drift

### **🎯 Funcionalidades Principais**
- Pipeline automatizado de ML
- Experiment tracking com MLflow
- Data versioning com DVC
- Fine-tuning de LLMs
- Sistemas RAG
- CI/CD para Machine Learning
- Monitoramento de modelos
- Deploy automatizado

---

### **📋 Pré-requisitos**
- Python 3.9+
- MLflow 2.8+
- DVC 3.0+
- Docker
- Git
- Storage (S3, GCS, Azure Blob)

### **🔧 Instalação Local**

```bash
# 1. Clone o repositório principal
git clone https://github.com/LucasBiason/ia-ml-knowledge-base.git
cd ia-ml-knowledge-base/projects/automated-ml-pipeline

# 2. Configure o ambiente virtual
python3.9 -m venv venv
source venv/bin/activate  # Linux/macOS
# ou: venv\Scripts\activate  # Windows

# 3. Instale as dependências
pip install -r requirements.txt

# 4. Configure DVC
dvc init
dvc remote add -d storage s3://your-bucket/ml-pipeline

# 5. Configure MLflow
export MLFLOW_TRACKING_URI=http://localhost:5000
mlflow ui --host 0.0.0.0 --port 5000

# 6. Execute a aplicação
uvicorn src.main:app --reload --host 0.0.0.0 --port 8003
```

### **🐳 Com Docker**

```bash
# Build da imagem
docker build -t automated-ml-pipeline .

# Executar container
docker run -p 8003:8003 -p 5000:5000 --env-file .env automated-ml-pipeline

# Ou com Docker Compose
docker-compose up -d
```

### **📊 Verificação da Instalação**

```bash
# Health check
curl http://localhost:8003/health

# MLflow UI
# Abra http://localhost:5000 no navegador

# Documentação da API
# Abra http://localhost:8003/docs no navegador

# Executar testes
pytest --cov=src --cov-report=term-missing
```

---

### **🔍 Health Check**

```bash
curl http://localhost:8003/health
```

**Resposta:**
```json
{
  "status": "healthy",
  "timestamp": "2025-01-XX",
  "version": "1.0.0",
  "mlflow_status": "connected",
  "dvc_status": "connected"
}
```

### **🔬 Criar Experimento**

```bash
curl -X POST "http://localhost:8003/api/v1/experiments/create" \
  -H "Content-Type: application/json" \
  -d '{
    "name": "sentiment_analysis_v1",
    "description": "Análise de sentimentos com BERT",
    "tags": {
      "model_type": "transformer",
      "task": "sentiment_analysis",
      "dataset": "twitter_sentiment"
    }
  }'
```

### **🤖 Fine-tuning de LLM**

```bash
curl -X POST "http://localhost:8003/api/v1/fine-tuning/start" \
  -H "Content-Type: application/json" \
  -d '{
    "model_name": "gpt-3.5-turbo",
    "task": "sentiment_analysis",
    "training_data": "data/training_data.jsonl",
    "hyperparameters": {
      "learning_rate": 1e-5,
      "batch_size": 4,
      "epochs": 3
    },
    "experiment_name": "sentiment_analysis_v1"
  }'
```

### **🔍 Sistema RAG**

```bash
curl -X POST "http://localhost:8003/api/v1/rag/query" \
  -H "Content-Type: application/json" \
  -d '{
    "query": "Como funciona o fine-tuning de LLMs?",
    "documents": ["docs/fine-tuning.md", "docs/llm-basics.md"],
    "model": "gpt-3.5-turbo",
    "max_tokens": 500
  }'
```

### **📊 Treinar Modelo**

```bash
curl -X POST "http://localhost:8003/api/v1/models/train" \
  -H "Content-Type: application/json" \
  -d '{
    "experiment_name": "sentiment_analysis_v1",
    "model_type": "transformer",
    "hyperparameters": {
      "model_name": "bert-base-uncased",
      "max_length": 512,
      "batch_size": 16,
      "epochs": 5,
      "learning_rate": 2e-5
    },
    "data_path": "data/sentiment_dataset.csv"
  }'
```

---

### **🎯 Progresso Atual**
- **📅 Início:** Semana 3 (18/08 - 24/08)
- **📈 Status:** Planejado
- **🎯 Entrega:** Final da Semana 3

### **✅ Funcionalidades Concluídas**
- [x] Estrutura base do projeto
- [x] Documentação e especificações
- [x] Padrões de qualidade definidos

### **🟡 Em Desenvolvimento**
- [ ] Setup do ambiente MLOps
- [ ] Configuração MLflow
- [ ] Configuração DVC
- [ ] Pipeline base de treinamento

### **⚪ Planejadas**
- [ ] Experiment tracking completo
- [ ] Fine-tuning de LLMs
- [ ] Sistemas RAG
- [ ] CI/CD para ML
- [ ] Monitoramento de modelos
- [ ] Deploy automatizado

---

### **🔧 Tecnologias Principais**
- **Python 3.9+** - Linguagem principal
- **FastAPI** - Framework web para APIs
- **MLflow** - Experiment tracking
- **DVC** - Data versioning
- **Docker** - Containerização

### **🤖 Machine Learning & IA**
- **scikit-learn** - Algoritmos de ML
- **TensorFlow/PyTorch** - Deep Learning
- **Transformers** - Modelos de NLP
- **LangChain** - Framework para LLMs
- **OpenAI API** - Fine-tuning e RAG

### **📊 MLOps & Monitoramento**
- **MLflow** - Experiment tracking
- **DVC** - Data versioning
- **Prometheus** - Métricas
- **Grafana** - Dashboards
- **Weights & Biases** - Experiment tracking (opcional)

### **🔍 Qualidade & Testes**
- **pytest** - Framework de testes
- **Black** - Formatação de código
- **isort** - Organização de imports
- **flake8** - Linting
- **mypy** - Type checking

---

## 📁 **ESTRUTURA DO PROJETO**

```
automated-ml-pipeline/
│
├── 📖 README.md                    # Este arquivo
├── 📋 requirements.txt             # Dependências Python
├── 🐳 Dockerfile                   # Configuração Docker
├── 🐳 docker-compose.yml          # Docker Compose
├── 🚫 .gitignore                   # Arquivos ignorados
├── ⚙️ .env.example                 # Variáveis de ambiente exemplo
├── 📄 .dvcignore                   # Arquivos ignorados pelo DVC
│
├── 🐍 src/                         # Código fonte
│   ├── 📄 main.py                 # Entry point FastAPI
│   ├── 🔧 config.py               # Configurações
│   ├── 📊 models.py               # Modelos Pydantic
│   │
│   ├── 🌐 api/                    # Rotas da API
│   │   ├── 📄 __init__.py
│   │   ├── 🔗 routes.py           # Rotas principais
│   │   ├── 🏥 health.py           # Health checks
│   │   ├── 🔬 experiments.py      # Gerenciamento de experimentos
│   │   ├── 🤖 fine_tuning.py      # Fine-tuning de LLMs
│   │   ├── 🔍 rag.py              # Sistemas RAG
│   │   ├── 📊 models.py           # Treinamento de modelos
│   │   └── 🚀 deploy.py           # Deploy de modelos
│   │
│   ├── 🧠 core/                   # Lógica de negócio
│   │   ├── 📄 __init__.py
│   │   ├── 🔬 experiment_manager.py # Gerenciamento de experimentos
│   │   ├── 🤖 fine_tuner.py       # Fine-tuning de LLMs
│   │   ├── 🔍 rag_system.py       # Sistemas RAG
│   │   ├── 📊 model_trainer.py    # Treinamento de modelos
│   │   ├── 🚀 model_deployer.py   # Deploy de modelos
│   │   └── 📈 model_monitor.py    # Monitoramento de modelos
│   │
│   ├── 🔄 pipeline/               # Pipeline MLOps
│   │   ├── 📄 __init__.py
│   │   ├── 🔬 data_pipeline.py    # Pipeline de dados
│   │   ├── 📊 training_pipeline.py # Pipeline de treinamento
│   │   ├── 🤖 evaluation_pipeline.py # Pipeline de avaliação
│   │   └── 🚀 deployment_pipeline.py # Pipeline de deploy
│   │
│   └── 🛠️ utils/                  # Utilitários
│       ├── 📄 __init__.py
│       ├── 🔧 mlflow_client.py    # Cliente MLflow
│       ├── 🔧 dvc_client.py       # Cliente DVC
│       ├── 📊 metrics.py          # Métricas
│       └── 🔧 storage.py          # Gerenciamento de storage
│
├── 🧪 tests/                       # Testes
│   ├── 📄 __init__.py
│   ├── 🧪 test_api.py             # Testes da API
│   ├── 🧪 test_experiments.py     # Testes de experimentos
│   ├── 🧪 test_fine_tuning.py     # Testes de fine-tuning
│   ├── 🧪 test_rag.py             # Testes de RAG
│   └── 🧪 test_pipeline.py        # Testes de pipeline
│
├── 📚 docs/                        # Documentação
│   ├── 📖 api.md                  # Documentação da API
│   ├── 🛠️ setup.md                # Guia de setup
│   ├── 💡 examples.md             # Exemplos de uso
│   ├── 🔬 experiments.md          # Guia de experimentos
│   └── 🤖 fine_tuning.md          # Guia de fine-tuning
│
├── 📊 data/                        # Dados
│   ├── 📄 raw/                    # Dados brutos
│   ├── 📄 processed/              # Dados processados
│   └── 📄 models/                 # Modelos treinados
│
├── 🔬 experiments/                 # Experimentos
│   ├── 📄 sentiment_analysis/     # Experimento 1
│   ├── 📄 text_classification/    # Experimento 2
│   └── 📄 fine_tuning/            # Experimento 3
│
└── 🐳 docker/                      # Configurações Docker
    ├── 📄 Dockerfile.dev          # Docker para desenvolvimento
    └── 📄 docker-compose.dev.yml  # Docker Compose para dev
```

---

### **⚙️ Variáveis de Ambiente**

```bash
# .env
APP_NAME=Automated-ML-Pipeline
APP_VERSION=1.0.0
DEBUG=True
HOST=0.0.0.0
PORT=8003

# MLflow
MLFLOW_TRACKING_URI=http://localhost:5000
MLFLOW_EXPERIMENT_NAME=default

# DVC
DVC_REMOTE=storage
DVC_STORAGE_PATH=s3://your-bucket/ml-pipeline

# OpenAI
OPENAI_API_KEY=your_openai_api_key
OPENAI_MODEL=gpt-3.5-turbo

# Storage
STORAGE_TYPE=s3
AWS_ACCESS_KEY_ID=your_access_key
AWS_SECRET_ACCESS_KEY=your_secret_key
AWS_DEFAULT_REGION=us-east-1

# Model Registry
MODEL_REGISTRY_URI=http://localhost:5000
```

### **🔬 Configuração MLflow**

```python
# src/config.py
from pydantic_settings import BaseSettings

class Settings(BaseSettings):
    app_name: str = "Automated-ML-Pipeline"
    app_version: str = "1.0.0"
    debug: bool = True
    
    # MLflow
    mlflow_tracking_uri: str = "http://localhost:5000"
    mlflow_experiment_name: str = "default"
    
    # DVC
    dvc_remote: str = "storage"
    dvc_storage_path: str = "s3://your-bucket/ml-pipeline"
    
    # OpenAI
    openai_api_key: str
    openai_model: str = "gpt-3.5-turbo"
    
    class Config:
        env_file = ".env"
```

---

### **🔧 Executar Testes**

```bash
# Todos os testes
pytest

# Com cobertura
pytest --cov=src --cov-report=term-missing

# Testes específicos
pytest tests/test_experiments.py -v

# Testes de pipeline
pytest tests/test_pipeline.py -v
```

### **📊 Cobertura de Testes**
- **Meta:** > 85% de cobertura
- **Foco:** Testes unitários, integração e pipeline
- **MLflow:** Testes específicos para experiment tracking

---

### **📊 Endpoints de Monitoramento**
- `GET /health` - Status da aplicação
- `GET /metrics` - Métricas Prometheus
- `GET /docs` - Documentação da API
- `GET /experiments` - Lista de experimentos

### **📈 Métricas Coletadas**
- Tempo de treinamento de modelos
- Performance de modelos (accuracy, precision, recall)
- Uso de recursos (CPU, GPU, memória)
- Drift de dados e modelos
- Tempo de deploy

---

### **🐳 Docker Production**

```bash
# Build para produção
docker build -t automated-ml-pipeline:prod .

# Executar em produção
docker run -d \
  --name automated-ml-pipeline \
  -p 8003:8003 \
  -p 5000:5000 \
  --env-file .env.prod \
  automated-ml-pipeline:prod
```

### **☁️ Cloud Deployment**
- **AWS:** ECS/Fargate com SageMaker
- **GCP:** Cloud Run com Vertex AI
- **Azure:** Container Instances com Azure ML

---

### **📝 Como Contribuir**
1. **Fork** o repositório principal
2. **Crie** uma branch para sua feature
3. **Desenvolva** seguindo os padrões de qualidade
4. **Teste** suas mudanças
5. **Abra** um Pull Request

### **🏆 Padrões de Contribuição**
- Siga os padrões definidos em `docs/padroes-qualidade.md`
- Mantenha cobertura de testes > 85%
- Use conventional commits
- Documente todas as mudanças

---

### **👤 Lucas Biason**
- **Email:** lucas.biason@foxcodesoftware.com
- **LinkedIn:** [Lucas Biason](https://www.linkedin.com/in/lucas-biason/)
- **GitHub:** [@LucasBiason](https://github.com/LucasBiason)

### **📚 Recursos Adicionais**
- **[Documentação Completa](../docs/)**
- **[Padrões de Qualidade](../docs/padroes-qualidade.md)**
- **[Cronograma Detalhado](../docs/cronograma-detailed.md)**

---

## 📄 **LICENÇA**

Este projeto está licenciado sob a Licença MIT - veja o arquivo [LICENSE](LICENSE) para detalhes.

---

### **📅 Semana 3 (18/08 - 24/08)**
- [ ] Setup completo do ambiente MLOps
- [ ] Configuração MLflow e DVC
- [ ] Pipeline base de treinamento
- [ ] Experiment tracking básico
- [ ] Fine-tuning de LLMs

### **🎯 Objetivos do Projeto**
- [ ] Sistemas RAG completos
- [ ] CI/CD para Machine Learning
- [ ] Monitoramento de modelos
- [ ] Deploy automatizado
- [ ] Documentação técnica

---

*Desenvolvido como parte do portfólio IA Knowledge Base por Lucas Biason.*
